src/generic/grid.cpp

```diff
@@ -5889,6 +5886,69 @@ void wxGrid::RefreshBlock(int topRow, int leftCol,
     }
 }
 
+void wxGrid::RefreshArea(int areas)
+{
+    if ( areas == wxGA_All )
+    {
+        // Normally if we want to refresh the entire grid we call
+        // Refresh() instead, but to make this function consistent
+        // and correct, just forward to wxScrolledCanvas::Refresh()
+        // if the argument is wxGA_All.
+        wxScrolledCanvas::Refresh();
+        return;
+    }
+
+    // Corner area
+    if ( (areas & wxGA_Corner) != 0 )
+        m_cornerLabelWin->Refresh();
+
+    int cw, ch;
+    GetClientSize(&cw, &ch);
+
+    // Cells area
+    if ( (areas & wxGA_Cells) != 0 )
+    {
+        const wxRect rect = wxRect(GetRowLabelSize(), GetColLabelSize(), cw, ch);
+
+        wxScrolledCanvas::Refresh(true, &rect);
+    }
+
+    // RowLabels area
+    if ( (areas & wxGA_RowLabels) != 0 && GetRowLabelSize() > 0 )
+    {
+        const wxRect rect = wxRect(0, GetColLabelSize(), GetRowLabelSize(), ch);
+
+        wxScrolledCanvas::Refresh(true, &rect);
+    }
+
+    // ColLabels area
+    if ( (areas & wxGA_ColLabels) != 0 && GetColLabelSize() > 0 )
+    {
+        const wxRect rect = wxRect(GetRowLabelSize(), 0, cw, GetColLabelSize());
+
+        wxScrolledCanvas::Refresh(true, &rect);
+    }
+}
+
+void wxGrid::RefreshArea(wxGridArea area, const wxRect& rect)
+{
+    wxASSERT_MSG( !rect.IsEmpty(), "RefreshArea expects a valid rectangle to refresh" );
+
+    if ( area == wxGA_Corner )
+    {
+        // Refreshing part of the corner area doesn't make sense, so refresh the entire
+        // corner area instead
+        m_cornerLabelWin->Refresh();
+    }
+
+    if ( area == wxGA_Cells )
+        wxScrolledCanvas::Refresh(true, &rect);
+    else if ( area == wxGA_RowLabels && GetRowLabelSize() > 0 )
+        wxScrolledCanvas::Refresh(true, &rect);
+    else if ( area == wxGA_ColLabels && GetColLabelSize() > 0 )
+        wxScrolledCanvas::Refresh(true, &rect);
```

Comment:
Written like this, it seems natural to ask ourselves if we really need to have this function? As it basically just does exactly the same thing as `Refresh()`, and only adds checks for having row/column labels. I think it would be simpler and clearer to just call `Refresh()` directly in the code.

Reply:
Actually the final version of the function is this:

```
void wxGrid::RefreshArea(wxGridArea area, wxRect& rect)
{
    wxASSERT_MSG( !rect.IsEmpty(), "RefreshArea expects a valid rectangle to refresh" );

    if ( area == wxGA_Cells )
    {
        rect.Offset(GetRowLabelSize(), GetColLabelSize());
    }
    else if ( area == wxGA_RowLabels && GetRowLabelSize() > 0 )
    {
        rect.x = 0;
        rect.width = m_rowLabelWidth;
        rect.Offset(0, GetColLabelSize());
    }
    else if ( area == wxGA_ColLabels && GetColLabelSize() > 0 )
    {
        rect.y = 0;
        rect.height = m_colLabelHeight;
        rect.Offset(GetRowLabelSize(), 0);
    }

    wxScrolledCanvas::Refresh(true, &rect);
}
```

i.e. we pass the rectangle by non-const reference and apply the invariants inside the function which will make the code more clearer (and robust?) at the call site.

And in fact, this function is called by SetCellValue(), SetRowLabelValue() and SetColLabelValue().
And will be used later (when we add transparent selection).

Reply:
Sorry, passing `rect` by non-const reference is _definitely_ a bad idea, nobody expects it to be modified inside this function, which looks like a `Refresh()` variant (and this one never modifies the rectangle passed to it), please let's not do this.

And I still don't understand the purpose of all this at all. E.g. for `wxGA_Cells` case the function seems to take a rectangle in `wxGridWindow` coordinates, which it then offsets by the grid window position and passes to `Refresh()`. How is this better than just calling `Refresh()` on `wxGridWindow` itself directly?

Reply:
> And I still don't understand the purpose of all this at all. E.g. for `wxGA_Cells` case the function seems to take a rectangle in `wxGridWindow` coordinates, which it then offsets by the grid window position and passes to `Refresh()`. How is this better than just calling `Refresh()` on `wxGridWindow` itself directly?

This was true when there were no frozen windows at all. But with the introduction of frozen windows, we have to stop thinking that the cell area is always a simple `wxGridWindow`.

Take for example how `wxGrid::DoSet{Row,Col}Size()` would become ridiculously simpler with `RefreshArea()`:
```
@@ -10085,7 +10123,17 @@ void wxGrid::DoSetRowSize( int row, int height )
 
         int y;
         CalcScrolledPosition(0, GetRowTop(topRow), NULL, &y);
+#if 1
+        int cw, ch;
+        GetClientSize(&cw, &ch);
 
+        wxRect updateRect(0, y, cw - GetRowLabelSize(), ch - y);
+        RefreshArea(wxGA_Cells, updateRect);
+
+        updateRect.y = y;
+        updateRect.height = ch - y;
+        RefreshArea(wxGA_RowLabels, updateRect);
+#else
         if ( topRow < m_numFrozenRows )
         {
             // This row is frozen, refresh the frozen windows.
@@ -10114,6 +10162,7 @@ void wxGrid::DoSetRowSize( int row, int height )
             if ( m_frozenColGridWin )
                 refreshLowerPart(m_frozenColGridWin);
         }
+#endif
     }
 }
 
@@ -10273,7 +10322,22 @@ void wxGrid::DoSetColSize( int col, int width )
 
         int x;
         CalcScrolledPosition(GetColLeft(leftCol), 0, &x, NULL);
+#if 1
+        int cw, ch;
+        GetClientSize(&cw, &ch);
 
+        wxRect updateRect(x, 0, cw - x, ch - GetColLabelSize());
+        RefreshArea(wxGA_Cells, updateRect);
+
+        // Refreshing the native header is unnecessary, as it updates
+        // itself correctly anyhow, and just results in extra flicker.
+        if ( !IsUsingNativeHeader() )
+        {
+            updateRect.x = x;
+            updateRect.width = cw - x;
+            RefreshArea(wxGA_ColLabels, updateRect);
+        }
+#else
         if ( leftCol < m_numFrozenCols )
         {
             FurtherWindowPartRefresher refreshFurtherPart(x);
@@ -10300,6 +10364,7 @@ void wxGrid::DoSetColSize( int col, int width )
             if ( m_frozenRowGridWin )
                 refreshFurtherPart(m_frozenRowGridWin);
         }
+#endif
     }
 }
```

Reply:
Could you please explain why can't we use `wxGrid::Refresh()` for doing this?

Reply:
Sure we can and is even more clear and concise.

First, remove the unnecessary block inside `wxGrid::Refresh()`

```
@@ -5707,7 +5704,7 @@ void wxGrid::Refresh(bool eraseb, const wxRect* rect)
     {
         // Refresh to get correct scrolled position:
         wxScrolledCanvas::Refresh(eraseb, rect);
-
+#if 0
         if (rect)
         {
             int rect_x, rect_y, rectWidth, rectHeight;
@@ -5798,6 +5795,7 @@ void wxGrid::Refresh(bool eraseb, const wxRect* rect)
             if ( m_frozenCornerGridWin )
                 m_frozenCornerGridWin->Refresh(eraseb, NULL);
         }
+#endif
     }
 }
``` 
because  `wxScrolledCanvas::Refresh(eraseb, rect)` will correctly refreshes the subwindows intersecting with the rectangle `rect`.

Now, the new `wxGrid::DoSet{Row,Col}Size()` versions look like:

```
@@ -10085,7 +10124,14 @@ void wxGrid::DoSetRowSize( int row, int height )
 
         int y;
         CalcScrolledPosition(0, GetRowTop(topRow), NULL, &y);
+        y += GetColLabelSize();
+#if 1
+        int cw, ch;
+        GetClientSize(&cw, &ch);
 
+        wxRect updateRect(0, y, cw, ch - y);
+        Refresh(true, &updateRect);
+#else
         if ( topRow < m_numFrozenRows )
         {
             // This row is frozen, refresh the frozen windows.
@@ -10114,6 +10160,7 @@ void wxGrid::DoSetRowSize( int row, int height )
             if ( m_frozenColGridWin )
                 refreshLowerPart(m_frozenColGridWin);
         }
+#endif
     }
 }
 
@@ -10273,7 +10320,14 @@ void wxGrid::DoSetColSize( int col, int width )
 
         int x;
         CalcScrolledPosition(GetColLeft(leftCol), 0, &x, NULL);
+        x += GetRowLabelSize();
+#if 1
+        int cw, ch;
+        GetClientSize(&cw, &ch);
 
+        wxRect updateRect(x, 0, cw - x, ch);
+        Refresh(true, &updateRect);
+#else
         if ( leftCol < m_numFrozenCols )
         {
             FurtherWindowPartRefresher refreshFurtherPart(x);
@@ -10300,6 +10354,7 @@ void wxGrid::DoSetColSize( int col, int width )
             if ( m_frozenRowGridWin )
                 refreshFurtherPart(m_frozenRowGridWin);
         }
+#endif
     }
 }
``` 
**The question now is:** what is the advantage of `RefreshArea()` (the one which takes `wxRect`) over just calling `wxGrid::Refresh()` directly ?
**The answer is:** It's more like a helper function that translates the rectangle into the `wxGrid` coordinate space (which `wxGrid::Refresh()` expects) more than anything else.

Reply:
IMHO this translation seems less clear than just doing it directly in the code, but if you think it's worth it, why not, but we don't really need the `area` parameter in this one at all then, do we? I.e. this coordinate translation is the same no matter what, isn't it?

Also, could you please document (in a comment before the function) which kind of coordinates exactly does it expect on entry?

Reply:
I'll do these then:
1. keep `RefreshArea(int areas)` function and get rid of the other overload then.
2. Simplify `wxGrid::Refresh()` and use it throughout the code when appropriate.
3. Document (in a comment) how to prepare the rectangle before passing it to `Refresh()`

What do you think ?

Reply:
A question: in the `wxGrid` implementation a call to `wxGrid::Refresh()` is guarded by a call to `ShouldRefresh()` which makes the second call to this function inside `Refresh()` redundant!
so is it ok/safe to remove the second call ? because I am not sure if user codes already call this function or not ?

Reply:
> I'll do these then:
> 
>     1. keep `RefreshArea(int areas)` function and get rid of the other overload then.
> 
>     2. Simplify `wxGrid::Refresh()` and use it throughout the code when appropriate.
> 
>     3. Document (in a comment) how to prepare the rectangle before passing it to `Refresh()`
> 
> 
> What do you think ?

Looks to me, thanks!

As for `ShouldRefresh()`, I think it's better to keep it in `Refresh()` because it's just a (probably helpful) internal optimization which doesn't cost much.